Pinecone: Add option to not store text

flash1293 opened the pull request:
This PR adds an option to the pinecone connector to control storing the embedded text as a metadata field in the index.
It's set to true by default, so it's no breaking change (the pydantic model treats a missing property as the default value).

This option is useful to reduce the required amount of storage in the destination and to avoid leaking confidential information.

Commit:
add option to not store text

Commit:
prepare release

flash1293 committed:
Automated Commit - Formatting Changes

Commit:
Merge remote-tracking branch 'origin/master' into flash1293/omit-text

Commit:
bump cdk version

flash1293 committed:
Automated Commit - Formatting Changes

Commit:
Merge branch 'master' into flash1293/omit-text

aaronsteers commented on airbyte-integrations/connectors/destination-pinecone/integration_tests/spec.json:
```diff
@@ -24,6 +24,12 @@
+          },
+          "store_text": {
+            "title": "Store text",
+            "description": "Store the text that gets embedded along with the vector and the metadata in the Pinecone index under the key `text`. If set to false, only the vector and the metadata will be stored.",
```
I'm not clear on the use case for this... Can you explain a bit more?

aaronsteers replied:
Thanks for your help to understand this paradigm. What do you think of this approach:

1. Invert the paradigm to "Don't store raw text" as an opt-in field, disabled by default.
2. Put this new setting behind "advanced" options if we can. (Perhaps we're already doing this but wanted to call it out.

Rationale would be:
1. This is really an edge case for purposes of large datasets and/or data privacy concerns.
2. Most users don't need to use this option and shouldn't use it if they are unsure.
3. If someone disables the text field, it will break the normal RAG use cases, in which the RAG flow expects to pull down the raw text as a result of the vector scan. 

I think phrasing in the inverse, default to disabled, slightly discourages this from being selected - unless the user knows they want this behavior and searches for it.

Wdyt?

I think also - while the nuance about whether stored in "metadata" specifically should probably be left out of the setting name. While it was helpful to me in better understanding the internals, it could be confusing to users who don't know the default behavior of Pinecone, such as when setting up for the first time. (We also have a separate "metadata" field input in our settings, and we don't want to conflate what we are _asking_ for as metadata, vs what the vector stores as metadata.

I think these nuances can be explained in the tooltip and in the docs (thanks for linking this field to them, btw 🙏). Lastly, I do think a generic "Don't store raw text" could be applicable to other destinations equally - and nicely abstract the internal implementation details for the purpose of accessibility.

Wdyt?

flash1293 replied:
Good points, I will close this PR and re-open as an extension of the vector DB CDK

aaronsteers commented:
Do we have any docs link for using Pinecone without storing text, and/or an issue where this is requested?

I'm still thinking about how to ensure users who want the recommended / default behavior not being confused by the extra option.

flash1293 commented:
It's actually the basic example to not include the text itself: https://docs.pinecone.io/docs/insert-data

I can add a section to the docs and auto-scroll to it when the user toggles this to explain in greater detail.

Commit:
improve documentation

Commit:
improve documentation

flash1293 commented:
@aaronsteers I added a section in the docs for this. If the user disables the setting, it will scroll to this section automatically:

https://github.com/airbytehq/airbyte/assets/1508364/efb2528e-0461-434e-bb11-08e37454c5c9


I also changed the label to "Store raw text as metadata". Do you think that's clear enough?

flash1293 commented:
Will move this logic to the vector DB CDK
